src/common/src/com/microsoftopentechnologies/tooling/msservices/helpers/auth/AADManagerImpl.java

```diff
@@ -0,0 +1,697 @@
+                            settings.getClientId());
+                    validateAuthenticationResult(authResult);
+
+                    UserInfo userInfo2 = new UserInfo(authResult.getUserInfo().getTenantId(),
+                            authResult.getUserInfo().getUniqueName());
+
+                    if (!userInfo.equals(userInfo2)) {
+                        throw new AzureCmdException("Invalid User Information retrieved");
+                    }
+
+                    setAuthenticationResult(userInfo, resource, authResult);
+                } finally {
+                    if (context != null) {
+                        context.dispose();
+                    }
+                }
+            }
+        } catch (Throwable t) {
+            if (isErrorTokenUnauthorized(t)) {
+                authenticateWithInteractiveToken(userInfo, resource, title);
+            } else if (t instanceof AzureCmdException) {
+                throw (AzureCmdException) t;
+            } else if (t instanceof ExecutionException) {
+                throw new AzureCmdException("Error invoking the authentication process with refresh token", t.getCause());
+            } else {
+                throw new AzureCmdException("Error invoking the authentication process with refresh token", t);
+            }
+        } finally {
+            // we release the lock before we do anything else as
+            // we don't want a lock leaking in case the statements
+            // following throw
+            userLock.writeLock().unlock();
+        }
+    }
+
+    private void authenticateWithInteractiveToken(@NotNull UserInfo userInfo,
+                                                  @NotNull String resource,
+                                                  @NotNull String title)
+            throws AzureCmdException {
+        AuthenticationResult authenticationResult = isAuthenticated(userInfo, resource) ?
+                getAuthenticationResult(userInfo, resource) :
+                null;
+        AuthenticationResult refreshAuthResult = hasRefreshAuthResult(userInfo) ?
+                getRefreshAuthResult(userInfo) :
+                null;
+        boolean useRenewedRefresh = false;
+
+        ReentrantReadWriteLock userLock = getTempUserLock(userInfo);
+        userLock.writeLock().lock();
+
+        try {
+            AuthenticationResult renewedAuth = isAuthenticated(userInfo, resource) ?
+                    getAuthenticationResult(userInfo, resource) :
+                    null;
+            AuthenticationResult renewedRefresh = hasRefreshAuthResult(userInfo) ?
+                    getRefreshAuthResult(userInfo) :
+                    null;
+            boolean useRenewedAuth = renewedAuth != null && !renewedAuth.equals(authenticationResult);
+            useRenewedRefresh = !useRenewedAuth && renewedRefresh != null && !renewedRefresh.equals(refreshAuthResult);
+
+            if (!useRenewedAuth && !useRenewedRefresh) {
```

Comment:
It might be worthwhile adding some documentation for the lines above in this method (`authenticateWithInteractiveToken`). From what I can tell we seem to be doing a double-checked lock above for authentication and presence of refresh tokens.